mmdet3d/models/dense_heads/fcaf3d_head.py

```diff
@@ -0,0 +1,682 @@
+            center_preds.append(center_pred)
+            bbox_preds.append(bbox_pred)
+            cls_preds.append(cls_pred)
+            points.append(point)
+        return center_preds[::-1], bbox_preds[::-1], cls_preds[::-1], \
+            points[::-1]
+
+    def forward_train(self, x, gt_bboxes, gt_labels, img_metas):
+        """Forward pass of the train stage.
+
+        Args:
+            x (list[SparseTensor]): Features from the backbone.
+            gt_bboxes (list[:obj:`BaseInstance3DBoxes`]): Ground truth
+                bboxes of each sample.
+            gt_labels(list[torch.Tensor]): Labels of each sample.
+            img_metas (list[dict]): Contains scene meta info for each sample.
+
+        Returns:
+            dict: Centerness, bbox and classification loss values.
+        """
+        center_preds, bbox_preds, cls_preds, points = self(x)
+        return self._loss(center_preds, bbox_preds, cls_preds, points,
+                          gt_bboxes, gt_labels, img_metas)
+
+    def forward_test(self, x, img_metas):
+        """Forward pass of the test stage.
+
+        Args:
+            x (list[SparseTensor]): Features from the backbone.
+            img_metas (list[dict]): Contains scene meta info for each sample.
+
+        Returns:
+            list[list[Tensor]]: bboxes, scores and labels for each sample.
+        """
+        center_preds, bbox_preds, cls_preds, points = self(x)
+        return self._get_bboxes(center_preds, bbox_preds, cls_preds, points,
+                                img_metas)
+
+    def _prune(self, x, scores):
+        """Prunes the tensor by score thresholding.
+
+        Args:
+            x (SparseTensor): Tensor to be pruned.
+            scores (SparseTensor): Scores for thresholding.
+
+        Returns:
+            SparseTensor: Pruned tensor.
+        """
+        with torch.no_grad():
+            coordinates = x.C.float()
+            interpolated_scores = scores.features_at_coordinates(coordinates)
+            prune_mask = interpolated_scores.new_zeros(
+                (len(interpolated_scores)), dtype=torch.bool)
+            for permutation in x.decomposition_permutations:
+                score = interpolated_scores[permutation]
+                mask = score.new_zeros((len(score)), dtype=torch.bool)
+                topk = min(len(score), self.pts_prune_threshold)
+                ids = torch.topk(score.squeeze(1), topk, sorted=False).indices
+                mask[ids] = True
+                prune_mask[permutation[mask]] = True
+        x = self.pruning(x, prune_mask)
+        return x
+
+    def _forward_single(self, x, scale):
+        """Forward pass per level.
+
+        Args:
+            x (SparseTensor): Per level neck output tensor.
+            scale (mmcv.cnn.Scale): Per level multiplication weight.
+
+        Returns:
+            tuple[Tensor]: Per level head predictions.
+        """
+        center_pred = self.center_conv(x).features
```

Comment:
We usually use conv_xxx to represent the last layers before the final results in mmdet3d convention and use xxx_convs to represent the shared convs before conv_xxx.
Should we use conv_center here?

Reply:
And please also check other layer names